lib/Conversion/MooreToCore/MooreToCore.cpp

```diff
@@ -169,142 +194,235 @@ struct ProcedureOpConversion : public OpConversionPattern<ProcedureOp> {
+    //
+    // - A "wait" block that reads the current state of any values used to
+    //   detect events and then waits until any of those values change. When a
+    //   change occurs, control transfers to the "check" block.
+    // - A "check" block which is executed after any interesting signal has
+    //   changed. This is where any `detect_event` ops read the current state of
+    //   interesting values and compare them against their state before the wait
+    //   in order to detect an event. If any events were detected, control
+    //   transfers to the "resume" block; otherwise control goes back to the
+    //   "wait" block.
+    // - A "resume" block which holds any ops after the `wait_event` op. This is
+    //   where control is expected to resume after an event has happened.
+    //
+    // Block structure before:
+    //     opA
+    //     moore.wait_event { ... }
+    //     opB
+    //
+    // Block structure after:
+    //     opA
+    //     cf.br ^wait
+    // ^wait:
+    //     <read "before" values>
+    //     llhd.wait ^check, ...
+    // ^check:
+    //     <read "after" values>
+    //     <detect edges>
+    //     cf.cond_br %event, ^resume, ^wait
+    // ^resume:
+    //     opB
+    auto *resumeBlock =
+        rewriter.splitBlock(op->getBlock(), ++Block::iterator(op));
+    auto *waitBlock = rewriter.createBlock(resumeBlock);
+    auto *checkBlock = rewriter.createBlock(resumeBlock);
+
+    auto loc = op.getLoc();
+    rewriter.setInsertionPoint(op);
+    rewriter.create<cf::BranchOp>(loc, waitBlock);
+
+    // We need to inline two copies of the `wait_event`'s body region: one is
+    // used to determine the values going into `detect_event` ops before the
+    // `llhd.wait`, and one will do the actual event detection after the
+    // `llhd.wait`.
+    //
+    // Create a copy of the entire `wait_event` op in the wait block, which also
+    // creates a copy of its region. Take note of all inputs to `detect_event`
+    // ops and delete the `detect_event` ops in this copy.
+    SmallVector<Value> valuesBefore;
+    rewriter.setInsertionPointToEnd(waitBlock);
+    auto clonedOp = cast<WaitEventOp>(rewriter.clone(*op));
+    for (auto detectOp :
+         llvm::make_early_inc_range(clonedOp.getOps<DetectEventOp>())) {
+      valuesBefore.push_back(detectOp.getInput());
+      rewriter.eraseOp(detectOp);
     }
 
-    rewriter.create<llhd::WaitOp>(loc, toObserve, Value(), ValueRange{}, check);
-    rewriter.setInsertionPointToStart(check);
+    // Determine the values used during event detection that are defined outside
+    // the `wait_event`'s body region. We want to wait for a change on these
+    // signals before we check if any interesting event happened.
+    SmallVector<Value> observeValues;
+    SmallDenseSet<Value> alreadyObserved;
+    clonedOp.getBody().walk([&](Operation *operation) {
+      for (auto value : operation->getOperands()) {
+        if (value.getParentRegion() == &clonedOp.getBody())
```

Comment:
If there are nested regions within the wait_event op region, we'd have to check if `clonedOp.getBody()` is an ancestor region of the value defining region.

Reply:
Nice catch! I've changed this and added a test for it.